### PR TITLE
add dlclassifier

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?
Close https://github.com/intel-analytics/BigDL/pull/1057 due to wrong git history and reopen it here.

primary changes:

Add DLClassifier and DLClassifierModel for the classification tasks, in which the prediction column is Double type.
Add unit tests for different feature and label types. Array[Double], Array[Float] and Vector.
Just pass RDD[Array[_]] to internalFit or internalTransform to avoid creating the extra array column.

## How was this patch tested?
added unit tests for DLEstimator and DLClassifier. Manual test for DLEstimatorLeNet example.

### CODE REVIEW

**Comment by hhbyyh** [comment]
cc @wzhongyuan

**Comment by hhbyyh** [comment]
Hi @wzhongyuan change the default maxEpoch to 100 to ensure better result on user data. Also I removed the hard accuracy check due to the data instability. I added RNG.setSeed but the trained model still is not stable.

**Comment by hhbyyh** [comment]
PR validation passed: http://172.168.2.101:8080/job/BigDL-PR-Validation/1150/

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 1]
```diff
@@ -15,34 +15,39 @@
...
+ * org.apache.spark.ml.linalg.{Vector, VectorUDT} for Spark 2.0+. Also label data can be of
+ * DoubleType.
+ * User should specify the feature data dimensions and label data dimensions via the constructor
+ * parameters featureSize and labelSize respectively. Internally the feature and label data are
+ * converted to BigDL tensors, to further train a BigDL model efficiently.
  *
- * For details usage, please refer to example :
- * [[com.intel.analytics.bigdl.example.MLPipeline.DLEstimatorLeNet]]
+ * For details usage, please refer to examples in package
+ * com.intel.analytics.bigdl.example.MLPipeline
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: How about adding a link [[...]] for it ?

Comment (hhbyyh) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: I'm afraid the package name may break the javadoc generation if some day we want to release the API doc.

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 2]
```diff
@@ -61,77 +66,61 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
...
-  protected override def internalFit(dataFrame: DataFrame): DLTransformerBase = {
-    val batches = toMiniBatch(dataFrame)
+  protected override def internalFit(
+      featureAndLabel: RDD[(Seq[AnyVal], Seq[AnyVal])]): DLTransformerBase = {
+    val batches = toMiniBatch(featureAndLabel)
     val dataset = DataSet.rdd(batches)
 
     val optimizer = Optimizer(model, dataset, criterion)
-      .setOptimMethod(new Adam[T]())
+      .setOptimMethod(new LBFGS[T]())
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Not sure if it's needed, but as maxEpoch, do you think we should provide interface for users to specify optim method as well ? Assuming for different tasks , different optim methods may be needed

Comment (hhbyyh) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Yes, that'll be in the following up PRs.

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 3]
```diff
@@ -147,3 +136,109 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
     copyValues(new DLEstimator(model, criterion, featureSize, labelSize), extra)
   }
 }
+
+
+/**
+ * [[DLModel]] helps embedding a BigDL model into a Spark Transformer, thus Spark users can
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: embedding -> embed ?

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala (hunk 2 above)
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Thanks, sounds great

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala (hunk 3 above)
Comment (hhbyyh) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Thanks, I'll change that. Have you finished the pass?

**Comment by wzhongyuan** [comment]
@hhbyyh   I went through all the changes, and I  saw you removed the redundant trait extending in DLClassifiler. As discussed in the thread , we will add another method for users to specify optim methods separately.
LGTM
Thanks

**Comment by hhbyyh** [comment]
Thanks @wzhongyuan

**Comment by hhbyyh** [comment]
Run the PR validation again for final check before merge.

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/job/BigDL-PR-Validation/ passed.

**Pull request merged**